### PR TITLE
consumer: cleanup RDY handling; fix pausing

### PR DESCRIPTION
**Pull request by mreiferson**
this addresses the issue reported in #222

### CODE REVIEW

**Comment by mreiferson** [comment]
RFR

File: mock_test.go [hunk 1]
```diff
@@ -111,25 +112,25 @@ func (n *mockNSQD) handle(conn net.Conn) {
...
+			n.t.Logf("mock: %s", line)
 			n.got = append(n.got, line)
 			params := bytes.Split(line, []byte(" "))
 			switch {
 			case bytes.Equal(params[0], []byte("IDENTIFY")):
 				l := make([]byte, 4)
 				_, err := io.ReadFull(rdr, l)
 				if err != nil {
-					log.Printf(err.Error())
+					n.t.Logf(err.Error())
```
Comment (ploxiln) [inline mock_test.go]: this should use `Log()` instead of `Logf()` (and 2 other cases below)

**Comment by ploxiln** [comment]
I noticed that `rdyRetryMtx` and `backoffMtx` can be plain Mutex instead of RWMutex

**Comment by ploxiln** [comment]
This is honestly the part of NSQ that I'm least familiar with ... the technical details of how ready-count is "refilled" and such ... I'll keep looking and slowly digesting though :)

File: mock_test.go (hunk 1 above)
Comment (mreiferson) [reply mock_test.go]: OK, fine

**Comment by mreiferson** [comment]
Addressed comments.

> the technical details of how ready-count is "refilled" and such

Yea, this is just old and outdated code that's already been [addressed in `pynsq`](https://github.com/nsqio/pynsq/pull/179). See https://github.com/nsqio/nsq/pull/404 for relevant context.

File: mock_test.go [hunk 2]
```diff
@@ -470,3 +469,73 @@ func TestConsumerBackoffDisconnect(t *testing.T) {
...
+	q.SetLogger(newTestLogger(t), LogLevelDebug)
+	q.AddHandler(&testHandler{})
+	err := q.ConnectToNSQD(n.tcpAddr.String())
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	timeoutCh := time.After(500 * time.Millisecond)
+	pauseCh := time.After(50 * time.Millisecond)
+	unpauseCh := time.After(75 * time.Millisecond)
```
Comment (mreiferson) [inline mock_test.go]: This is a little fragile and time dependent, but should be reasonably stable.

File: conn.go [hunk 3]
```diff
@@ -207,13 +206,12 @@ func (c *Conn) RDY() int64 {
 
 // LastRDY returns the previously set RDY count
```
Comment (ploxiln) [inline conn.go]: I guess this function is somewhat deprecated now?

File: mock_test.go [hunk 4]
```diff
@@ -470,3 +469,73 @@ func TestConsumerBackoffDisconnect(t *testing.T) {
 		}
 	}
 }
+
+func TestConsumerPause(t *testing.T) {
+	msgIDGood := MessageID{'1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 'a', 's', 'd', 'f', 'g', 'h'}
```
Comment (ploxiln) [inline mock_test.go]: not an expected message id, as it has characters that are not hex ... not sure if it matters at all

very minor idea, alternate way to write this:

```go
	var msgId MessageID
	copy(msgId[:], []byte("0123456789abcdef"))
```

File: mock_test.go [hunk 5]
```diff
@@ -470,3 +469,73 @@ func TestConsumerBackoffDisconnect(t *testing.T) {
...
+
+	msgGood := NewMessage(msgIDGood, []byte("good"))
+
+	script := []instruction{
+		// SUB
+		instruction{0, FrameTypeResponse, []byte("OK")},
+		// IDENTIFY
```
Comment (ploxiln) [inline mock_test.go]: I think IDENTIFY is before SUB?

**Comment by ploxiln** [comment]
looks good 👍

File: conn.go (hunk 3 above)
Comment (mreiferson) [reply conn.go]: Yea

File: mock_test.go (hunk 4 above)
Comment (mreiferson) [reply mock_test.go]: copied from previous tests and yes, it doesn't matter

File: mock_test.go (hunk 5 above)
Comment (mreiferson) [reply mock_test.go]: Fixing

**Pull request merged**